### ISSUE TITLE
Add single table inheritance (STI) feature.

### ISSUE THREAD

**Issue by trevorpe** [vuex-orm/vuex-orm#310]
This doesn't seem possible currently, but is there any plan (or even ability: I'm not a JavaScript guy) to have inheritance mapping similar to [Doctrine's single-table inheritance](https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/inheritance-mapping.html#single-table-inheritance)?

I have a case where I have multiple types of tasks, each a subclass of `Task`. It seems that the polymorphic relationships that are currently implemented are for cases where multiple parent types may own `Task` objects. I have the opposite problem, where I have several types of tasks and need to have a parent entity own many (potentially differently typed) tasks.

I've seen issues #50 and #192, but they both require a sub-key rather than taking advantage of inheritance. While this will work, it requires the creation of a separate entity when all of the required discriminator information is already present.

It seems to me that it would be possible to have an inheritance mapping and factory method to properly instantiate the correct model type, but I'm not familiar with the code base or the normalizr library. What are your thoughts? I'm happy to help if there's anything I can do.

**Comment by rkingon** [comment]
We have a similar situation.

Say we have a base model `Loan` and two inheritance models `PersonalLoan` `AutoLoan` that both `extend` `Loan`. `User` model `hasMany` of this inheritance collection.

(trying to use a primitive example).

An API payload could look something like:
```javascript
User {
  ...
  loans: [
    { id, type: 'PERSONAL', amount, purpose },
    { id, type: 'AUTO', amount, year, make, model },
    { id, type: 'AUTO', amount, year, make, model }
  ]
}
```

Upon digestion of the API payload, it would be nice to assign the class based on `type` -- so if `type === 'PERSONAL'` the `PersonalLoan` class is used, etc.

Is this possible in the current implementation? I know there are potentially a few work arounds by introducing another model which facilitates the relationship, but as @trevorpe mentioned, it seems unnecessary (and in my experience, that too can become very buggy). 

Have to give you kudos @kiaking, this is one solid orm!

**Comment by rkingon** [comment]
I forked an existing Code Sandbox to help illustrate. https://codesandbox.io/s/ppk0lplz2j

What we have here is a `User` that `hasMany` `Loans` -- We have additional models `PersonalLoan` `AutoLoan` which extend `Loan` -- trying to find a way to tie the `Loans` to their respective models based on the `type` attribute.

The reason is that we may have very different class methods, computed properties, etc across the `types`

**Comment by rkingon** [comment]
As a test, I tried to "nest" the "types"

Here is a codesandbox: https://codesandbox.io/s/k9q5w2wmj7

I've changed the DTO a bit to have "specific" fields nested within the parent, so essentially:
```
{
  id,
  type,
  sharedFields,
  typeNestedModel: {
    specificFields
  }
}
```

it's not my favorite approach, but would work... 

sorry to blow up this thread -- just something we're desperately working on!

**Comment by kiaking** [comment]
Hi! Sorry for my late reply. Maybe I'm not getting your requirement, but to extend the model, I think you could do this?

```js
import { Model } from "@vuex-orm/core";
import User from "./user";

export default class Loan extends Model {
  static entity = "loans";
  static primaryKey = "id";

  static fields() {
    return {
      id: this.increment(),
      type: this.attr(),
      amount: this.attr(),
      userId: this.attr(),
      user: this.belongsTo(User, "userId")
    };
  }
}


import Loan from "./loan";

export default class PersonalLoan extends Loan {
  static entity = "personalLoans";

  static fields() {
    return {
      ...super.fields(),
      purpose: this.attr()
    };
  }
}
```

This is going to create a fields that combines both parent and it self.

**Comment by trevorpe** [comment]
Thanks for the reply, @kiaking.

I think the feature both I and @rkingon were referring to would be handling the polymorphism of this inheritance tree *in relationships* (i.e. the inverse side of that loan-to-user relationship). Currently, it seems that one must create a separate entity. Essentially, the difference is that `type` key refers to the *current entity*, not another through a relationship.

For example, if the `User` entity needs to have a `hasMany` relationship to `Loan`, one would normally do the following:
```
import { Model } from "@vuex-orm/core";
import Loan from "./loan";

export default class User extends Model {
  static entity = "users";

  static fields() {
    return {
      id: this.increment(),
      // Name, email, etc.
      loans: this.hasMany(Loan, 'userId')
    };
  }
}
```
But if there are many subclasses of `Loan`, it doesn't look like you will get instances of the subclass here, just the fields available in the base class (untested, correct me if I'm wrong). I was hoping to be able to have an array of mixed loan types (the type specified by the `type` key on the `Loan` entity itself). Currently, one would have to create another entity to hold the subclass-specific fields and relate them through a `morphTo` relationship (like #50 and #192).

**Comment by rkingon** [comment]
@trevorpe 💯

**Comment by kiaking** [comment]
@trevorpe Thank you for the clarification! I think I understand now.

This is an interesting idea. I have no experience with Doctrine but never done the same in ORM like Eloquent or Active Record.

Currently, all relationship refers to a single entity (which in this case would be an intermediate table which has `morhTo` capability). I think technically it is possible to refer to multiple entities. Not sure how the definition API would look like but something like following?

```js
export default class User extends Model {
  static entity = "users"

  static fields() {
    return {
      id: this.increment(),
      loans: this.hasManyMorph(Loan, 'userId', 'type')
    }
  }
}
```

It is surely an interesting idea and I'm not against implementing this type of relation. It might open up a new way of dealing data at the front end.

Do you know any ORM implementation for this relationship? Like the name of the relation and how they define the relationship?

**Comment by rkingon** [comment]
@kiaking I am surprised you have not used this in Active Record -- it is where I learned it from!

https://api.rubyonrails.org/classes/ActiveRecord/Inheritance.html

**Comment by rkingon** [comment]
@kiaking I think ideally, you would not specify the `type` on the relationship level, but rather in the model.

Very similar to `fields()` we can have `types()` where you build a map of types. The `key` is the possible value of the attribute `type` and the `value` is the `Class`

```js
const loans = [
  { id, type: 'PERSONAL', etc }
]

static types() {
    return {
      PERSONAL: PersonalLoanModel,
      AUTO: AutoLoanModel
    }
  }
```

This way you can map out the expected types. By simply passing `type` as an argument, you wouldn't necessarily know "how to map it" to the class. Also, you shouldn't need to rinse/repeat for anything that could `hasMany`

**Comment by kiaking** [comment]
Thank you for the reference! Well I'm too Laravel guy I guess: )

This is pretty cool and I'm willing to go for it. But first, I need to focus on fixing other bugs and publishing v1.0.0, but I would like to keep this issue open as an enhancement. Meanwhile, maybe someone could tackle this issue.

**Comment by rkingon** [comment]
I took a quick peek and realized the project is in TypeScript. I've yet to dabble in TS, so I might do more harm than good, unfortunately.

For example purposes, here is STI in Laravel: https://github.com/tightenco/parental

**Comment by kiaking** [comment]
@rkingon Thanks for the reference! I really like the idea so I'll make sure to come back to this. While doing so, maybe it would be cool if we can start discussion on designing the API. Like where and how to declare `static types`, and how everything should work and such.

**Comment by trevorpe** [comment]
I've not dabbled in TypeScript either, but I might have some free time later this week or next week to take a peek under the hood.

Until I dig into a bit more code, I can't really speak to anything too low-level, but from a high-level I see a few decisions/considerations to keep in mind:
1. **How to define the discriminator mapping?**
I like @rkingon's idea of a static `types()` method.
1. **How to define the discriminator key?**
Could be a new field type, or maybe just a static `discriminator` property that returns the key to use as the discriminator.
```js
import { Model } from "@vuex-orm/core";
import PersonalLoan from "./personal-loan";
import AutoLoan from "./auto-loan";

export default class Loan extends Model {
  static entity = "loans";
  // Static property idea
  static discriminator = "loanType";

  static types() {
    return {
      PERSONAL: PersonalLoan,
      AUTO: AutoLoan,
    };
  }

  static fields() {
    return {
      id: this.increment(),
      // Or field type idea
      loanType: this.discriminator(),
    };
  }
}
```
3. **Should this polymorphic entity work in the existing relationships?**
Take my previous example (which is pretty well what I envisioned when opening the issue):
```js
import { Model } from "@vuex-orm/core";
import Loan from "./loan";

export default class User extends Model {
  static entity = "users";

  static fields() {
    return {
      id: this.increment(),
      // Other entity has foreign key
      loans: this.hasMany(Loan, 'userId'),

      // Current entity has foreign key
      loanIds: this.attr(null),
      otherLoans: this.hasManyBy(Loan, 'loanIds')
    };
  }
}
```
Should the existing relationships handle polymorphic entities (identifiable by when `types()` returns a discriminator mapping), or will all new relationships have to be created specifically to handle inheritance (mainly to signal to the ORM that subclasses must be checked as well):
```js
import { Model } from "@vuex-orm/core";
import Loan from "./loan";

export default class User extends Model {
  static entity = "users";

  static fields() {
    return {
      id: this.increment(),
      // Other entity has foreign key
      loans: this.hasManyMorph(Loan, 'userId'),

      // Current entity has foreign key
      loanIds: this.attr(null),
      otherLoans: this.hasManyMorphBy(Loan, 'loanIds')
    };
  }
}
```
4. **Will it be "true" single-table inheritance mapping?**
That is, will each subclass still define its own static `entity` property and be stored in a separate "table" in the Vuex store, or will they all be stored under a single "table?" I would assume if they are stored in the same table, it will require some reworking of existing code since querying through a subclass should not match instances of the parent or sibling classes.

**Comment by rkingon** [comment]
To chime in on @trevorpe points

**How to define the discriminator mapping?**
Sounds like we're all in agreement here

**How to define the discriminator key?**
I think it could default to `type`, overridable by a static getter, ie: `inheritanceColumn` (that is the name rails uses: https://api.rubyonrails.org/classes/ActiveRecord/ModelSchema/ClassMethods.html#method-i-inheritance_column-3D -- doesn't make it right, just what I defaulted to)

**Should this polymorphic entity work in the existing relationships?**
`hasMany` `hasManyBy` etc should not be "affected" imo -- all should work as it does currently -- the only difference is that if the instances may end up being SubClass instances. Ie:
```js
// [ {id, type: 'PERSONAL'}, {id, type: 'AUTO'} ]
static fields() {
  id: this.attr(),
  loanIds: this.attr([]),
  loans: this.hasManyBy(Loan, 'loanIds')
}

// userInstance.loans output
// [ PersonalLoanInstance, AutoLoanInstance ]
```

**Will it be "true" single-table inheritance mapping?**
This one is tough, since there is also a concept of Multi Table Inheritance (which is arguably better, depending on the situation -- ie: are fields mostly shared or mostly different).

using the current API, I would assume both to return results
`getters['entities/loans/query']` and `getters['entities/personalLoans/query']`
`loans` would return "all types" whereas `personalLoans` would return only `loans` with `type=PERSONAL`

^^ that definitely adds complexity though!

**Comment by kiaking** [comment]
I mostly agree with you guys as well.

**How to define the discriminator key?**  
I like the idea of `type` as well. It sounds more related to `types`. But maybe we could be bit more explicit and go with `typeKey` like `primaryKey`.

```js
import { Model } from '@vuex-orm/core'
import PersonalLoan from './PersonalLoan'
import AutoLoan from './AutoLoan'

export default class Loan extends Model {
  static entity = 'loans'

  static typeKey = 'loanType'

  static types () {
    return {
      PERSONAL: PersonalLoan,
      AUTO: AutoLoan
    }
  }

  static fields () {
    return {
      id: this.increment(),
      loanType: this.string(''),
    }
  }
}
```

**Should this polymorphic entity work in the existing relationships?**
I agree with @rkingon. If a user is defining STI, then the user should be expecting to get all entities ending up being corresponding instances.

**Will it be "true" single-table inheritance mapping?**  
I would say it should be stored in single table. The big difference between Vuex ORM and RDB is that Vuex ORM actually doesn't have any columns since Vuex Store is just an JS object.

Technically, we can store any object with completely different schema. This might be useful when defining STI. So inside store it could look like this.

```js
{
  loans: {
    data: {
      '1': {
        id: 1,
        loanType: 'PERSONAL'
      },
      '2': {
        id: 2,
        loanType: 'AUTO',
        year: 2015
      }
    }
  }
}
```

I'm not sure if this is the way to go, I mean if all record should have same field with null for the field that specific type doesn't own, or we should keep it simple as above.

> using the current API, I would assume both to return results
> getters['entities/loans/query'] and getters['entities/personalLoans/query']
> loans would return "all types" whereas personalLoans would return only loans with type=PERSONAL

Agree. This sounds fair. Well, even if we split the table, we have to check and do some special logic for TSI models so I think having all records inside single table is slightly easier to implement interns of fetching and querying data. As @rkingon mentioned, it is going to be a nightmare if we have to join separate tables when calling getters['entities/loans/query'] 🤔

**Comment by trevorpe** [comment]
I think the idea of defaulting to `type`, but allowing one to override it using `typeKey` makes sense.

I also think that existing relationships should be used and behave as expected. Entities without polymorphism would behave as they do now, and "new" polymorphic entities would have the proper instances be returned depending on the value of the type key.

I agree that having everything in a single place is easiest to deal with. There is new logic to be written no matter how it's structured, but I suspect that many of the arguments for multiple table inheritance don't really apply when everything is stored in memory and when you're working in JS where there is no requirement to have the same "columns" and types. Separating them would just make for a nightmare of trying to stitch objects together.

**Comment by rkingon** [comment]
+1 on `typeKey` as it is consistent with `primaryKey`

**Comment by tvillaren** [comment]
Hello, 

Has anyone started working / fiddling on this?   
Just encountered the same issue and might give it a try in the next days. 
I've been reading all comments and the design proposed sounds good to me!

**Comment by trevorpe** [comment]
I'm using the work-around for now and it's working just fine, and I don't really have the time to commit to much, so if you'd like to tackle this, feel free.

I'm happy to help where I can, I just have higher priority items that need my focus, so I can't really commit to learning enough about the code base to implement this feature.

**Comment by tvillaren** [comment]
I know the feeling :) 
What do you call "the workaround" ? 

Currently, if I have different types of Loan coming from an API, I was unable to create the corresponding instances in the state (i.e if I want to do an `insert` of the "base Loan" entity but with an array of derived Loans, it will not create the derived ones accordingly. Am I right?

I need to organise my thoughts as well on this and finish a couple of other things, but might give it a try in the coming days.

**Comment by trevorpe** [comment]
The workaround is described in #50 and #192, but it is pretty well as follows:
```js
import {Model} from '@vuex-orm/core';

/**
 * Loan class
 */
export default class Loan extends Model {
  static entity = 'loans';

  static fields () {
    return {
      id: this.number(null),
      loanType: this.string('autoLoans'),
      content: this.morphTo('id', 'loanType'),
    }
  }
}
```

```js
import {Model} from '@vuex-orm/core';

/**
 * AutoLoan class
 */
export default class AutoLoan extends Model {
  static entity = 'autoLoans';

  static fields() {
    return {
      id: this.number(null),
      make: this.attr(''),
      model: this.attr(''),
    }
  }
}

```
With the work-around, the base class has to define some sort of field to hold the fields on the sub-types and refer to it through a `morphTo` relationship, keeping the same IDs between both.

What we were discussing was the idea of having single-table inheritance, where the `Loan` class can be sub-classed and have a type key to determine which model type to use. The information is all there anyway, so it is more elegant to have a parent class and sub-classes rather than making another entity just to hold the fields that differ between sub-classes.

**Comment by pymarco** [comment]
Hi, I just wanted to add that I found this ticket when searching to see if the orm implemented a feature similar to Django's generic foreignkeys. It is a similar use case, but perhaps a little different. In this scenario, STI isn't relevant but the ability for relate a model to any other model by specifying a type field. A general overview can be [found here](https://simpleisbetterthancomplex.com/tutorial/2016/10/13/how-to-use-generic-relations.html). Doing this on the server side, however, [has it's drawbacks](https://lukeplant.me.uk/blog/posts/avoid-django-genericforeignkey/).

I have recently played around with implementing a notifications system that draws from the [Activity Streams Spec](http://activitystrea.ms/specs/atom/1.0/). For example a notification model contains fields:
* ``Actor``. The object that performed the activity.
* ``Verb``. The verb phrase that identifies the action of the activity.
* ``Action Object``. *(Optional)* The object linked to the action itself.
* ``Target``. *(Optional)* The object to which the activity was performed.

where `actor`, `action_object`, and `target` would be generic foreignkeys. I learned this design [from this library](https://github.com/django-notifications/django-notifications/blob/master/README.rst#django-notifications-documentation), but prefer to implement the server side model with uuid and type fields that the orm can later resolve to their various frontend models.

I'm not in a position to contribute toward implementing this, but wanted to share the scenario in case it is found useful to the project.

As a side, I've been reading through the Vuex-ORM issues the last few days as I'm curious to begin integrating it into my project. @kiaking I find the positivity in your comments very inspiring. Thank you for your work on this project.

**Comment by tvillaren** [comment]
@pymarco I'm not sure this has to do with inheritance, does it? It makes me think of this issue: https://github.com/vuex-orm/vuex-orm/issues/379 

All, PR 384 (https://github.com/vuex-orm/vuex-orm/pull/384) is implementing STI (use of discriminator field is optional). Documentation is provided as well.

**Comment by pymarco** [comment]
Thanks @tvillaren , my original comment was off topic. I've since found the documentation for polymorphic relations. Cheers.

**Comment by kiaking** [comment]
OK the PR is merged! Just need few more things done before making a new release. Stay tuned forks!

**Comment by kiaking** [comment]
Now Single Table Inheritance (STI) is available at Vuex ORM 0.32.0! 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.32.0

Huge thanks to @tvillaren for wonderful PR. And thank you all for contributing to the design discussion. Maybe there'll be things to improve the feature. Looking forward for your feedbacks! 🙌
